imageloading-core/src/main/java/com/google/accompanist/imageloading/AsyncImage.kt

```diff
@@ -0,0 +1,366 @@
+    shouldRefetchOnSizeChange: (currentResult: ImageLoadState, size: IntSize) -> Boolean,
+    content: @Composable BoxScope.(imageLoadState: ImageLoadState) -> Unit
+) {
+    if (LocalInspectionMode.current && previewPlaceholder != 0) {
+        // If we're in inspection mode (preview) and we have a preview placeholder, just draw
+        // that using an Image and return
+        Image(
+            painter = painterResource(previewPlaceholder),
+            contentDescription = null,
+            modifier = modifier,
+        )
+        return
+    }
+
+    val coroutineScope = rememberCoroutineScope()
+
+    val imageLoader = remember(request, coroutineScope) {
+        ImageLoader(
+            state = request,
+            coroutineScope = coroutineScope,
+            shouldRefetchOnSizeChange = shouldRefetchOnSizeChange,
+        )
+    }
+
+    // NOTE: All of the things that we want to be able to change without recreating the whole object
+    // we want to do inside of here.
+    SideEffect {
+        imageLoader.shouldRefetchOnSizeChange = shouldRefetchOnSizeChange
+    }
+
+    Box(
+        propagateMinConstraints = true,
+        modifier = modifier.then(imageLoader.layoutModifier)
+    ) {
+        content(request.loadState)
+    }
+}
+
+@Stable
+private class ImageLoader<R : Any>(
+    val state: AsyncImageState<R>,
+    val coroutineScope: CoroutineScope,
+    shouldRefetchOnSizeChange: (currentResult: ImageLoadState, size: IntSize) -> Boolean,
+) : RememberObserver {
+
+    // the size of the image, as informed by the layout system and the request.
+    //
+    // This value will be read during layout
+    private var requestSize by mutableStateOf<IntSize?>(null)
+
+    var colorFilter by mutableStateOf<ColorFilter?>(null)
+    var contentScale by mutableStateOf(ContentScale.Fit)
+    var alignment by mutableStateOf(Alignment.Center)
+    var alpha by mutableStateOf(1f)
+    var shouldRefetchOnSizeChange by mutableStateOf(shouldRefetchOnSizeChange)
+
+    // Current request job
+    private var job: Job? = null
+
+    val layoutModifier = Modifier.layout { measurable, constraints ->
+        // NOTE: this is where the interesting logic is, but there shouldn't be anything here that
+        // you can't do that you're doing using BoxWithConstraints currently.
+        val newSize = IntSize(
+            width = if (constraints.hasBoundedWidth) constraints.maxWidth else -1,
+            height = if (constraints.hasBoundedHeight) constraints.maxHeight else -1
+        )
+
+        if (requestSize == null ||
+            (requestSize != newSize && shouldRefetchOnSizeChange(state.loadState, newSize))
+        ) {
+            requestSize = newSize
+        }
+
+        val placeable = measurable.measure(constraints)
+        layout(width = placeable.width, height = placeable.height) {
+            placeable.place(0, 0)
+        }
+    }
+
+    val paintModifier = object : PainterModifier() {
```

Comment:
Why not `Modifier.painter` ?

Reply:
Because `Modifier.paint` would mean that we need to useecomposition to recreate the modifier when any of the values changes. This subclass allows us to do state reads _in_ the modifier

Reply:
Added comments